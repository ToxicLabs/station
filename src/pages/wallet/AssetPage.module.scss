@import "mixins";

.details {
  @include flex(center);
  flex-direction: column;

  padding: 1.8rem 1rem;
  padding-top: 2.6rem;
  height: fit-content;
  background-color: var(--card-bg);
  grid-area: details;

  h1 {
    font-size: 20pt;
    margin-top: 0.6rem;
  }

  p {
    color: var(--text-muted);
    font-size: 12pt;
    margin-top: 0.2rem;
  }
}

.chainlist {
  background-color: var(--card-bg-muted);
  padding-top: 1.6rem;
  padding-inline: 10px;
  grid-area: list;
  display: grid;
  grid-template-areas: "title" "list";
  grid-template-rows: min-content auto;
  overflow-y: hidden;

  h3 {
    padding-left: 10px;
    grid-area: title;
    margin-bottom: 10px;
  }
}

.chainlist__list {
  overflow-y: auto;
  grid-area: list;
  padding-bottom: 1.6rem;
}

.asset_info_denom {
  @include flex(flex-start);
  align-items: center;
  width: fit-content;
  background-color: var(--button-default-bg);
  margin-top: 1rem;
  padding: 0.2rem 0.6rem;
  border-radius: 2rem;
  font-size: 10pt;
  font-weight: 500;
}

.actions {
  @include flex(space-between);
  background-color: var(--card-bg);
  padding: 1rem;
  gap: 0.6rem;

  button {
    min-width: 0;
    width: 100%;
    padding-block: 20px;
    height: fit-content;
  }
}
