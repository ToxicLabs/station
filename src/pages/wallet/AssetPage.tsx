import { useNativeDenoms } from "data/token"
import { useWalletRoute, Path } from "./Wallet"
import styles from "./AssetPage.module.scss"
import { Read, TokenIcon } from "components/token"
import { useCurrency } from "data/settings/Currency"
import { useExchangeRates } from "data/queries/coingecko"
import { useBankBalance } from "data/queries/bank"
import AssetChain from "./AssetChain"
import { Button, CopyIcon } from "components/general"
import { useTranslation } from "react-i18next"
import { capitalize } from "@mui/material"
import Vesting from "./Vesting"
import { isTerraChain } from "utils/chain"
import { Tag } from "components/display"

const AssetPage = () => {
  const currency = useCurrency()
  const { data: prices } = useExchangeRates()
  const balances = useBankBalance()
  const readNativeDenom = useNativeDenoms()
  const { t } = useTranslation()
  const { setRoute, route } = useWalletRoute()
  const denom = route.path === Path.coin ? route.denom : "uluna"
  const { token, symbol, icon, decimals } = readNativeDenom(denom)

  const filteredBalances = balances.filter(
    (b) => readNativeDenom(b.denom).token === token
  )
  const totalBalance = filteredBalances.reduce(
    (acc, b) => acc + parseInt(b.amount),
    0
  )
  const price = symbol?.endsWith("...") ? 0 : prices?.[token]?.price ?? 0

  return (
    <>
      <section className={styles.details}>
        <TokenIcon token={token} icon={icon} size={50} />
        <div className={styles.asset_info_denom}>
          <Tag color={"default"}>
            <span>
              {denom.length > 12
                ? denom.slice(0, 5) +
                  "..." +
                  denom.slice(denom.length - 5, denom.length)
                : denom}
            </span>
          </Tag>

          <span style={{ marginTop: 4 }}>
            <CopyIcon text={denom} />
          </span>
        </div>
        <h1>
          {currency.symbol}{" "}
          <Read
            decimals={decimals}
            amount={totalBalance * price}
            fixed={2}
            token={symbol}
          />
        </h1>
        <p>
          <Read decimals={decimals} amount={totalBalance} token={symbol} />
          {symbol}
        </p>
      </section>
      <section className={styles.chainlist}>
        <h3>{t("Chains")}</h3>
        <div className={styles.chainlist__list}>
          {filteredBalances
            .sort((a, b) => parseInt(b.amount) - parseInt(a.amount))
            .map((b, i) => (
              <div key={i}>
                <AssetChain
                  symbol={symbol}
                  balance={b.amount}
                  chain={b.chain}
                  token={token}
                  decimals={decimals}
                />
                {token === "uluna" && isTerraChain(b.chain) && <Vesting />}
              </div>
            ))}
        </div>
      </section>
      <section className={styles.actions}>
        <Button
          color="primary"
          onClick={() =>
            setRoute({
              path: Path.send,
              denom,
              previousPage: route,
            })
          }
        >
          {t("Send")}
        </Button>
        <Button
          onClick={() =>
            setRoute({
              path: Path.receive,
              previousPage: route,
            })
          }
        >
          {capitalize(t("receive"))}
        </Button>
      </section>
    </>
  )
}

export default AssetPage
