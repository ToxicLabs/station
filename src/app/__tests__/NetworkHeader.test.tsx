import React from "react"
import NetworkHeader from "../sections/NetworkHeader"
import { render } from "@testing-library/react"
import { useNetworkName } from "../../data/wallet"

function renderComponent() {
  return render(<NetworkHeader />)
}

jest.mock("../../data/wallet", () => ({
  useNetworkName: jest.fn(),
}))

describe("NetworkHeader component matches snapshots", () => {
  it("matches component mainnet", () => {
    useNetworkName.mockReturnValueOnce("mainnet")
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })

  it("matches component testnet", () => {
    useNetworkName.mockReturnValueOnce("testnet")
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })

  it("matches component localterra", () => {
    useNetworkName.mockReturnValueOnce("localterra")
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })

  it("matches component classic", () => {
    useNetworkName.mockReturnValueOnce("classic")
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
