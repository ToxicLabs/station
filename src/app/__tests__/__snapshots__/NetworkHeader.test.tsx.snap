// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkHeader component matches snapshots matches component classic 1`] = `
<DocumentFragment>
  <div
    class="component"
  >
    CLASSIC
  </div>
</DocumentFragment>
`;

exports[`NetworkHeader component matches snapshots matches component localterra 1`] = `
<DocumentFragment>
  <div
    class="component"
  >
    LOCALTERRA
  </div>
</DocumentFragment>
`;

exports[`NetworkHeader component matches snapshots matches component mainnet 1`] = `<DocumentFragment />`;

exports[`NetworkHeader component matches snapshots matches component testnet 1`] = `
<DocumentFragment>
  <div
    class="component"
  >
    TESTNET
  </div>
</DocumentFragment>
`;
